content/things/js-to-asc.md

```diff
@@ -0,0 +1,563 @@
+
+### Growth
+
+After doing some [profiling with d8] on a build with debugging symbols (`--debug`), it turns out that a lot of time is spent in a system library `libsystem_platform.dylib`, which contains OS-level primitives for threading and memory management. Calls into this library are made from `__new` and `__renew`, which in turn are called from `Array<f32>#push`:
+```
+[Bottom up (heavy) profile]:
+  ticks parent  name
+  18670   96.1%  /usr/lib/system/libsystem_platform.dylib
+  13530   72.5%    Function: *~lib/rt/itcms/__renew
+  13530  100.0%      Function: *~lib/array/ensureSize
+  13530  100.0%        Function: *~lib/array/Array<f32>#push
+  13530  100.0%          Function: *binaryheap_optimized/BinaryHeap<f32>#push
+  13530  100.0%            Function: *binaryheap_optimized/push
+   5119   27.4%    Function: *~lib/rt/itcms/__new
+   5119  100.0%      Function: *~lib/rt/itcms/__renew
+   5119  100.0%        Function: *~lib/array/ensureSize
+   5119  100.0%          Function: *~lib/array/Array<f32>#push
+   5119  100.0%            Function: *binaryheap_optimized/BinaryHeap<f32>#push
+```
+
+Clearly, we have a problem with allocations here. But JavaScript somehow manages to make an ever-growing array fast, so why can’t AssemblyScript? Luckily, the standard library of AssemblyScript is rather small and approachable, so let’s go and [take a look][array push impl] at this ominous `push()` function of the `Array<T>` class:
+
+```ts
+export class Array<T> {
+  // ...
+  push(value: T): i32 {
+    var length = this.length_;
+    var newLength = length + 1;
+    ensureSize(changetype<usize>(this), newLength, alignof<T>());
+    // ...
+    return newLength;
+  }
+  // ...
+}
+```
+
+The `push()` function correctly determines that the new length of the array is the current length plus 1 and then calls `ensureSize()`, to make sure that the underlying buffer has enough room (“capacity”) to grow to this length.
+
+```ts
+function ensureSize(array: usize, minSize: usize, alignLog2: u32): void {
+  // ...
+  if (minSize > <usize>oldCapacity >>> alignLog2) {
+    // ...
+    let newCapacity = minSize << alignLog2;
+    let newData = __renew(oldData, newCapacity);
+    // ... 
+  }
+}
+```
+
+`ensureSize()`, in turn, checks if the capacity is smaller than the new `minSize`, and if so, allocates a new buffer of size  `minSize` using `__renew`, which entails copying all the data from the old buffer to the new buffer. For that reason our benchmark, where we push _one million values_ one-by-one into the array, ends up causing a _lot_ of allocation work and create a lot of garbage. 
+
+In other languages, like [Rust’s `std::vec`][rust impl] or [Go’s slices][go impl], the new buffer has _double_ the old buffer’s capacity, which amortizes the allocation work over time. [I am working to fix this in ASC][asc issue], but in the meantime we can create our own `CustomArrow<T>` that has the desired behavior. Lo and behold, we made things faster!
```

Comment:
FWIW, when V8 needs to grow a JavaScript array's backing store, it adds 50% (which is the same idea as doubling, just trading a few more grow operations for less wasted memory, but still avoiding the terrible complexity of growing by 1 element at a time -- the key is that the growth step is a fraction of the (existing or required) capacity, not an additive constant).